pkg/datapath/loader/compile.go

```diff
@@ -0,0 +1,182 @@
+			Source:     endpointProg,
+			Output:     endpointAsm,
+			OutputType: outputAssembly,
+		},
+		{
+			Source:     endpointProg,
+			Output:     endpointProg,
+			OutputType: outputSource,
+		},
+	}
+	datapathProg = &progInfo{
+		Source:     endpointProg,
+		Output:     endpointObj,
+		OutputType: outputObject,
+	}
+)
+
+// irPath determines a temporary path on the filesystem to store the compiled
+// intermediate representation object for the program.
+//
+// TODO: Remove this and just pipe the 'clang' and 'llc' programs together.
+func irPath(prog *progInfo, dir *directoryInfo) string {
+	return fmt.Sprintf("%s/%s.bc", dir.Output, prog.Output)
+}
+
+// progLDFlags determines the loader flags for the specified prog and paths.
+func progLDFlags(prog *progInfo, dir *directoryInfo) []string {
+	return []string{
+		fmt.Sprintf("-filetype=%s", prog.OutputType),
+		irPath(prog, dir),
+		"-o", fmt.Sprintf("%s/%s", dir.Output, prog.Output),
+	}
+}
+
+// Link the specified program from the specified path to the intermediate
+// representation, to the output specified in the prog's info.
+func Link(ctx context.Context, prog *progInfo, dir *directoryInfo, debug bool) error {
+	args := make([]string, 0, 8)
+	if debug {
+		args = append(args, "-mattr=dwarfris")
+	}
+	args = append(args, standardLDFlags...)
+	args = append(args, progLDFlags(prog, dir)...)
+
+	out, err := exec.CommandContext(ctx, linker, args...).CombinedOutput()
+	if err != nil {
+		scanner := bufio.NewScanner(bytes.NewReader(out))
+		return fmt.Errorf("Failed to link %q: %s: %q", prog.Output, err, scanner.Text())
+	}
+
+	return nil
+}
+
+// progLDFlags determines the compiler flags for the specified prog and paths.
+func progCFlags(prog *progInfo, dir *directoryInfo) []string {
+	var output string
+
+	if prog.OutputType == outputSource {
+		output = fmt.Sprintf("%s/%s", dir.Output, prog.Output)
+	} else {
+		output = irPath(prog, dir)
+	}
+
+	return []string{
+		fmt.Sprintf("-I%s/globals", dir.Runtime),
+		fmt.Sprintf("-I%s", dir.Output),
+		fmt.Sprintf("-I%s/include", dir.Library),
+		"-c", fmt.Sprintf("%s/%s", dir.Library, prog.Source),
+		"-o", output,
+	}
+}
+
+// compile and link a program.
+func compile(ctx context.Context, prog *progInfo, dir *directoryInfo, debug bool) error {
+	args := make([]string, 0, 16)
+	if prog.OutputType == outputSource {
+		args = append(args, "-E") // Preprocessor
+	} else {
+		args = append(args, "-emit-llvm")
+		if debug {
+			args = append(args, "-g")
+		}
+	}
+	args = append(args, standardCFlags...)
+	args = append(args, progCFlags(prog, dir)...)
+
+	// Compilation is split between two exec calls. First clang generates
+	// LLVM bitcode and then later llc compiles it to assembly.
+	out, err := exec.CommandContext(ctx, compiler, args...).CombinedOutput()
+	if err != nil {
+		scanner := bufio.NewScanner(bytes.NewReader(out))
+		return fmt.Errorf("Failed to compile %q: %s: %q", prog.Output, err, scanner.Text())
+	}
+	defer os.Remove(irPath(prog, dir))
```

Comment:
If line 165 fails, does it mean the path won't be clean up?

Reply:
If line 165 fails, then the path won't be created in the first place. Clang creates it, we pass it into llc, then clean up afterwards.